File: bindings/ruby/drtrace.rb
# drtrace.rb — in-process native runtime tracing for Ruby, backed by DynamoRIO.
#
# This is the Ruby counterpart to bindings/python/asmtest/drtrace.py: the optional
# DynamoRIO native-trace tier (see include/asmtest_drtrace.h and
# docs/native-tracing.md). Where the emulator tier (Asmtest::Emu) traces isolated
# guest bytes, NativeTrace traces host-native code as it runs **inside this Ruby
# process**: bring DynamoRIO up once, materialize host-native machine code, mark a
# region, call into it, and read back basic-block coverage / the instruction stream.
#
# Like asmtest.rb it keeps all Fiddle FFI inside; unlike asmtest.rb it loads a
# *separate* shared object — libasmtest_drapp — which dlopen()s libdynamorio lazily
# on the C side once the client is configured, so nothing here links DynamoRIO.
#
# Advanced, Linux-x86-64-only, and opt-in. NativeTrace.available? reports whether the
# tier can run (built + libdynamorio resolvable) so callers self-skip cleanly. The
# lib path comes from env ASMTEST_DRAPP_LIB, else the gem-bundled slot
# native/<os>-<cpu>/libasmtest_drapp.<ext> next to this file, else
# <repo>/build/libasmtest_drapp.so; a load failure is caught so available? returns
# false rather than raising.
#
# Example:
#   require_relative "drtrace"
#   Asmtest::DrTrace::NativeTrace.start(client: "./build/libasmtest_drclient.so",
#                                       dynamorio_home: "/opt/DynamoRIO")
#   code = Asmtest::DrTrace::NativeCode.from_bytes(
#            [0x48, 0x89, 0xF8, 0x48, 0x01, 0xF0, 0xC3].pack("C*"))  # mov rax,rdi; add; ret
#   tr = Asmtest::DrTrace::NativeTrace.create(blocks: 64, instructions: 0)
#   tr.register("add", code)
#   tr.region("add") { @r = code.call(20, 22) }
#   raise unless @r == 42 && tr.covered?(0)
#   Asmtest::DrTrace::NativeTrace.shutdown
require "fiddle"

module Asmtest
  # The optional DynamoRIO native-trace tier. Namespaced apart from the core
  # Asmtest module so requiring it is explicit (it loads a different .so).
  module DrTrace
    # Status code shared by the lifecycle + registration calls (ASMTEST_DR_OK).
    OK = 0

    VOIDP = Fiddle::TYPE_VOIDP
    LONG  = Fiddle::TYPE_LONG
    INT   = Fiddle::TYPE_INT
    SZ    = Fiddle::TYPE_SIZE_T
    LL    = Fiddle::TYPE_LONG_LONG
    VOID  = Fiddle::TYPE_VOID

    # The gem-bundled native slot next to this file (how the gem ships payloads),
    # mirroring Asmtest.resolve_emu_lib in asmtest.rb: native/<os>-<cpu>/.
    def self.bundled_slot
      os  = RbConfig::CONFIG["host_os"] =~ /darwin/ ? "darwin" : "linux"
      cpu = RbConfig::CONFIG["host_cpu"] =~ /arm|aarch64/ ? "arm64" : "x86_64"
      ext = os == "darwin" ? "dylib" : "so"
      [os, cpu, ext]
    end

    # The absolute path libasmtest_drapp actually resolved to (set by
    # resolve_drapp_lib), for library_path below.
    @resolved_path = nil

    # Resolve libasmtest_drapp. Order: an explicit ASMTEST_DRAPP_LIB wins (dev /
    # custom build); else the gem-bundled slot native/<os>-<cpu>/ next to this file
    # (how the gem ships it); else <repo>/build/libasmtest_drapp.so (the in-tree
    # build artifact). The build/ fallback stays BELOW the bundled path so an
    # installed gem never prefers a leaked checkout.
    def self.resolve_drapp_lib
      env = ENV["ASMTEST_DRAPP_LIB"].to_s
      unless env.empty?
        @resolved_path = env
        return env
      end
      os, cpu, ext = bundled_slot
      bundled = File.expand_path("native/#{os}-#{cpu}/libasmtest_drapp.#{ext}", __dir__)
      if File.exist?(bundled)
        @resolved_path = bundled
        return bundled
      end
      repo = File.expand_path("../../build/libasmtest_drapp.so", __dir__)
      @resolved_path = repo
      repo
    end

    # The gem-bundled DR client shipped alongside libasmtest_drapp, if present:
    # $ASMTEST_DRCLIENT wins, else native/<os>-<cpu>/libasmtest_drclient.<ext> next
    # to this file, else the <repo>/build tree; nil when none exists (the C side
    # then takes its own $ASMTEST_DRCLIENT / repo-build fallback). Mirrors
    # drtrace.py's _default_client.
    def self.default_client
      env = ENV["ASMTEST_DRCLIENT"].to_s
      return env unless env.empty?
      os, cpu, ext = bundled_slot
      cands = [
        File.expand_path("native/#{os}-#{cpu}/libasmtest_drclient.#{ext}", __dir__),
        File.expand_path("../../build/libasmtest_drclient.so", __dir__),
      ]
      cands.find { |c| File.exist?(c) }
    end

    # The absolute path of the libasmtest_drapp this process resolved (resolving it
    # if it hasn't been yet). Lets a clean-room test assert the bundled tier — not a
    # leaked build/ tree — satisfied the load. Mirrors drtrace.py's library_path().
    def self.library_path
      resolve_drapp_lib if @resolved_path.nil?
      @resolved_path
    end

    def self.func(lib, name, args, ret)
      Fiddle::Function.new(lib[name], args, ret)
    end

    # Load the lib and bind every entry point, OR leave LIB/FN nil if it can't load
    # (the tier requires DynamoRIO and may be absent). available? folds both the
    # load failure and a runtime asmtest_dr_available()==0 into a clean false.
    begin
      LIB = Fiddle.dlopen(resolve_drapp_lib)
      FN = {
        # ---- process-wide lifecycle ----
        available:    func(LIB, "asmtest_dr_available", [], INT),
        init:         func(LIB, "asmtest_dr_init", [VOIDP], INT),
        start:        func(LIB, "asmtest_dr_start", [], INT),
        stop:         func(LIB, "asmtest_dr_stop", [], INT),
        shutdown:     func(LIB, "asmtest_dr_shutdown", [], VOID),
        marker_error: func(LIB, "asmtest_dr_marker_error", [], INT),
        # ---- region registration + markers ----
        register:     func(LIB, "asmtest_dr_register_region", [VOIDP, VOIDP, SZ, VOIDP], INT),
        register_symbol: func(LIB, "asmtest_dr_register_symbol", [VOIDP, SZ, VOIDP], INT),
        symbol_demo:  func(LIB, "asmtest_symbol_demo", [LONG, LONG], LONG),
        unregister:   func(LIB, "asmtest_dr_unregister_region", [VOIDP], INT),
        trace_begin:  func(LIB, "asmtest_trace_begin", [VOIDP], VOID),
        trace_end:    func(LIB, "asmtest_trace_end", [VOIDP], VOID),
        # ---- host-native executable code ----
        exec_alloc:   func(LIB, "asmtest_exec_alloc", [VOIDP, SZ, VOIDP], INT),
        exec_free:    func(LIB, "asmtest_exec_free", [VOIDP], VOID),
        # ---- the app-owned trace handle + accessors (from the shared trace.o) ----
        trace_new:    func(LIB, "asmtest_trace_new", [SZ, SZ], VOIDP),
        trace_free:   func(LIB, "asmtest_trace_free", [VOIDP], VOID),
        trace_covered: func(LIB, "asmtest_trace_covered", [VOIDP, LL], INT),
        blocks_len:   func(LIB, "asmtest_emu_trace_blocks_len", [VOIDP], LL),
        insns_total:  func(LIB, "asmtest_emu_trace_insns_total", [VOIDP], LL),
        insns_len:    func(LIB, "asmtest_emu_trace_insns_len", [VOIDP], LL),
        block_at:     func(LIB, "asmtest_emu_trace_block_at", [VOIDP, SZ], LL),
        insn_at:      func(LIB, "asmtest_emu_trace_insn_at", [VOIDP, SZ], LL),
      }.freeze
    rescue Fiddle::DLError
      LIB = nil
      FN = nil
    end

    # Host-native machine code in real executable (W^X) memory. Mirrors the C
    # asmtest_exec_code_t {void* base; size_t len;}: 16 bytes, base at offset 0.
    class NativeCode
      EXEC_CODE_SIZE = 16

      # Wrap an already-allocated out-struct pointer (a Fiddle::Pointer over the
      # 16-byte asmtest_exec_code_t the alloc wrote into).
      def initialize(ec_ptr)
        @ec = ec_ptr
        @base = @ec[0, 8].unpack1("Q")   # void* base  (offset 0)
        @len = @ec[8, 8].unpack1("Q")    # size_t len  (offset 8)
        @freed = false
      end

      # Map executable memory and copy +bytes+ (a binary String) into it, returning
      # a NativeCode whose #base is the live entry address.
      def self.from_bytes(bytes)
        ptr = Fiddle::Pointer.new(Fiddle.malloc(EXEC_CODE_SIZE), EXEC_CODE_SIZE)
        ptr[0, EXEC_CODE_SIZE] = "\x00".b * EXEC_CODE_SIZE
        rc = DrTrace::FN[:exec_alloc].call(bytes, bytes.bytesize, ptr)
        if rc != DrTrace::OK
          Fiddle.free(ptr.to_i)
          raise "asmtest_exec_alloc failed: #{rc}"
        end
        new(ptr)
      end

      # Entry address of the executable mapping (offset 0 = entry).
      def base
        @base
      end

      # Number of code bytes copied.
      def length
        @len
      end

      # Invoke the code through a function pointer with +a+ and +b+ as C longs,
      # reading the result back as a long (the SysV integer ABI). Fiddle::Function
      # accepts the raw entry address as its first argument.
      def call(a = 0, b = 0)
        Fiddle::Function.new(@base, [LONG, LONG], LONG).call(a, b)
      end

      # Unmap the executable memory. The caller MUST unregister the range first if it
      # was registered (see the C header) — NativeTrace#unregister does that.
      def free
        return if @freed
        DrTrace::FN[:exec_free].call(@ec)
        Fiddle.free(@ec.to_i)
        @freed = true
      end
    end

    # An app-owned coverage recorder for a registered native region. Wraps the opaque
    # asmtest_trace_t handle; instruction recording when instructions > 0, block
    # recording when blocks > 0.
    class NativeTrace
      attr_reader :handle

      def initialize(handle)
        @handle = handle
      end

      # ---- process-wide lifecycle (class-level; one DynamoRIO per process) ----

      # True if the tier can run: the lib loaded AND libdynamorio is resolvable.
      # Folds a load failure (FN nil) and a runtime asmtest_dr_available()==0 into
      # one clean false, so callers self-skip without a rescue.
      def self.available?
        return false if DrTrace::FN.nil?
        DrTrace::FN[:available].call != 0
      end

      # Bring DynamoRIO up in-process and take over (init then start). +client+ is the
      # path to libasmtest_drclient.so (else DrTrace.default_client: $ASMTEST_DRCLIENT,
      # then the gem-bundled slot / repo build; else the C side's own fallback);
      # +dynamorio_home+ lets the C side find libdynamorio (else $ASMTEST_DR_LIB /
      # rpath); +client_options+ are extra client options; +mode+ is the process-init
      # default recording mode. Raises on a nonzero rc.
      #
      # Named #start (not Ruby's reserved #initialize, and distinct from the instance
      # region begin) to read as a lifecycle verb at the class level.
      def self.start(client: nil, dynamorio_home: nil, client_options: nil, mode: 0)
        client ||= DrTrace.default_client # prefer the bundled DR client in a gem
        opts = DrTrace.build_options(client, dynamorio_home, client_options, mode)
        rc = DrTrace::FN[:init].call(opts)
        raise "asmtest_dr_init failed: #{rc}" if rc != DrTrace::OK
        rc = DrTrace::FN[:start].call
        raise "asmtest_dr_start failed: #{rc}" if rc != DrTrace::OK
      end

      # Stop DynamoRIO and clean up (dr_app_stop_and_cleanup), returning to UNINIT.
      def self.shutdown
        DrTrace::FN[:shutdown].call
      end

      # Count of illegal marker operations since init (0 == every marker balanced).
      def self.marker_error
        DrTrace::FN[:marker_error].call
      end

      # Call the exported asmtest_symbol_demo fixture (a*2+b) that symbol-mode
      # tests trace by name.
      def self.symbol_demo(a, b)
        DrTrace::FN[:symbol_demo].call(a, b)
      end

      # ---- per-trace ----

      # Allocate a trace handle. NOTE the C order: asmtest_trace_new(insns_cap,
      # blocks_cap) — instructions FIRST, blocks SECOND. Named #create (not #new)
      # because the handle comes from C, not a plain Ruby allocation.
      def self.create(blocks: 64, instructions: 0)
        h = DrTrace::FN[:trace_new].call(instructions, blocks)
        raise "asmtest_trace_new failed" if h.null?
        new(h)
      end

      # Register a non-overlapping native range under +name+, recording coverage into
      # this trace. The client resolves the range; +code+ is a NativeCode.
      def register(name, code)
        rc = DrTrace::FN[:register].call(name, code.base, code.length, @handle)
        raise "register_region(#{name.inspect}) failed: #{rc}" if rc != DrTrace::OK
        self
      end

      # Symbol mode: trace a named exported function with no begin/end markers —
      # recording is always-on over [entry, entry+max_len). The client resolves
      # +name+ across all loaded modules.
      def register_symbol(name, max_len = 256)
        rc = DrTrace::FN[:register_symbol].call(name, max_len, @handle)
        raise "register_symbol(#{name.inspect}) failed: #{rc}" if rc != DrTrace::OK
        self
      end

      # Drop the named region (the client drops its cached translation). Call this
      # before NativeCode#free per the C header's unregister-then-free ordering.
      def unregister(name)
        DrTrace::FN[:unregister].call(name)
      end

      # Open recording for +name+, run the block, then close it — markers stay
      # balanced even if the block raises (the ensure runs end). Returns the block's
      # value.
      def region(name)
        DrTrace::FN[:trace_begin].call(name)
        begin
          yield
        ensure
          DrTrace::FN[:trace_end].call(name)
        end
      end

      # True if the basic block at byte-offset +off+ (from the region entry) was entered.
      def covered?(off)
        DrTrace::FN[:trace_covered].call(@handle, off) != 0
      end

      # Number of distinct basic blocks recorded.
      def blocks_len
        DrTrace::FN[:blocks_len].call(@handle)
      end

      # Total instructions in the ordered instruction stream (instruction mode).
      def insns_total
        DrTrace::FN[:insns_total].call(@handle)
      end

      # The distinct basic-block start offsets recorded, in first-seen order.
      def block_offsets
        n = DrTrace::FN[:blocks_len].call(@handle)
        (0...n).map { |i| DrTrace::FN[:block_at].call(@handle, i) }
      end

      # The ordered instruction-offset stream actually stored — each executed
      # instruction's offset in execution order, up to the trace's insns capacity
      # (insns_len, not the possibly-larger insns_total).
      def insn_offsets
        n = DrTrace::FN[:insns_len].call(@handle)
        (0...n).map { |i| DrTrace::FN[:insn_at].call(@handle, i) }
      end

      def free
        return unless @handle
        DrTrace::FN[:trace_free].call(@handle)
        @handle = nil
      end
    end

    # Build the asmtest_drtrace_options_t the C side reads: three const char* then an
    # int mode. The layout is {ptr, ptr, ptr, int} = 3 * 8 + 4, padded to 32 bytes on
    # x86-64. A nil/empty string becomes a NULL pointer so the C side takes its env
    # fallback (e.g. $ASMTEST_DRCLIENT for the client). The string buffers and the
    # struct are kept alive on the returned pointer so the GC can't free them mid-call.
    def self.build_options(client, dynamorio_home, client_options, mode)
      opts = Fiddle::Pointer.new(Fiddle.malloc(32), 32)
      opts[0, 32] = "\x00".b * 32
      held = []
      write_strp = lambda do |off, s|
        next if s.nil? || s.empty?
        buf = Fiddle::Pointer[s + "\x00"]   # NUL-terminated C string, GC-pinned below
        held << buf
        opts[off, 8] = [buf.to_i].pack("Q")
      end
      write_strp.call(0, dynamorio_home) # const char* dynamorio_home
      write_strp.call(8, client)         # const char* client_path
      write_strp.call(16, client_options) # const char* client_options
      opts[24, 4] = [mode].pack("l")     # asmtest_drtrace_mode_t mode (int)
      opts.instance_variable_set(:@asmtest_held, held) # pin string buffers to opts
      opts
    end
  end
end

File: bindings/lua/drtrace.lua
-- drtrace.lua — in-process native runtime tracing for LuaJIT, backed by DynamoRIO.
--
-- This is the Lua mirror of the Python `asmtest.drtrace` surface (see
-- bindings/python/asmtest/drtrace.py and include/asmtest_drtrace.h). Where the
-- emulator tier (asmtest.lua's Trace) traces isolated guest bytes, `NativeTrace`
-- traces host-native code as it runs **inside this LuaJIT process**: initialize
-- DynamoRIO once, materialize host-native machine code, mark a region, call into
-- it, and read back basic-block coverage / the instruction stream.
--
-- It loads libasmtest_drapp and drives the C API; libdynamorio is dlopen()ed
-- lazily by the C side after the client is configured, so nothing here links DR.
-- The tier is advanced, Linux-x86-64-only, and opt-in: NativeTrace.available()
-- reports whether it can run so callers self-skip cleanly. The load itself is
-- wrapped in pcall, so a missing libasmtest_drapp (no DynamoRIO) self-skips too.
--
-- The shared library is resolved in the same order as asmtest.lua's core loader:
--   ASMTEST_DRAPP_LIB   an explicit path wins (dev / custom build)
--   native/<os>-<arch>/ the rock-bundled slot next to this file (published rock)
--   <repo>/build/...     the in-tree build artifact (below the bundled path)
local ffi = require("ffi")

ffi.cdef([[
typedef struct { const char* dynamorio_home; const char* client_path; const char* client_options; int mode; } asmtest_drtrace_options_t;
typedef struct { void* base; size_t len; } asmtest_exec_code_t;
int  asmtest_dr_available(void);
int  asmtest_dr_init(const asmtest_drtrace_options_t*);
int  asmtest_dr_start(void);
int  asmtest_dr_stop(void);
void asmtest_dr_shutdown(void);
int  asmtest_dr_register_region(const char* name, void* base, size_t len, void* trace);
int  asmtest_dr_register_symbol(const char* symbol, size_t max_len, void* trace);
int  asmtest_dr_unregister_region(const char* name);
long asmtest_symbol_demo(long a, long b);
void asmtest_trace_begin(const char* name);
void asmtest_trace_end(const char* name);
int  asmtest_dr_marker_error(void);
int  asmtest_exec_alloc(const uint8_t* bytes, size_t len, asmtest_exec_code_t* out);
void asmtest_exec_free(asmtest_exec_code_t*);
void* asmtest_trace_new(size_t insns_cap, size_t blocks_cap);
void  asmtest_trace_free(void*);
int   asmtest_trace_covered(void* trace, uint64_t off);
uint64_t asmtest_emu_trace_blocks_len(void* trace);
uint64_t asmtest_emu_trace_insns_total(void* trace);
uint64_t asmtest_emu_trace_insns_len(void* trace);
uint64_t asmtest_emu_trace_block_at(void* trace, size_t i);
uint64_t asmtest_emu_trace_insn_at(void* trace, size_t i);
]])

local ASMTEST_DR_OK = 0

-- This module's own directory, so the rock-bundled slot + repo build/ resolve
-- relative to the file (mirrors asmtest.lua). This file lives at bindings/lua/,
-- so the repo root is two directories up.
local MODULE_DIR = (debug.getinfo(1, "S").source:sub(2):match("(.*/)")) or "./"

-- The rock-bundled native slot (os, arch, ext), mirroring asmtest.lua's core
-- loader: native/<os>-<arch>/lib*.<ext> next to this file (how the rock ships it).
local function bundled_slot()
  local os_name = ffi.os == "OSX" and "darwin" or "linux"
  local arch = ffi.arch == "arm64" and "arm64" or "x86_64" -- LuaJIT 'x64' -> x86_64
  local ext = ffi.os == "OSX" and "dylib" or "so"
  return os_name, arch, ext
end

-- The absolute path libasmtest_drapp actually resolved to (set by drapp_path),
-- for M.library_path below.
local resolved_path = nil

-- Resolve libasmtest_drapp. Order: an explicit ASMTEST_DRAPP_LIB wins (dev /
-- custom build); else the rock-bundled slot native/<os>-<arch>/ next to this file
-- (how the rock ships it); else <repo>/build/ (the in-tree build artifact). The
-- build/ fallback stays BELOW the bundled path so an installed rock never prefers
-- a leaked checkout.
local function drapp_path()
  local p = os.getenv("ASMTEST_DRAPP_LIB")
  if p and p ~= "" then resolved_path = p; return p end
  local os_name, arch, ext = bundled_slot()
  local bundled = MODULE_DIR .. "native/" .. os_name .. "-" .. arch .. "/libasmtest_drapp." .. ext
  local f = io.open(bundled, "r")
  if f then f:close(); resolved_path = bundled; return bundled end
  local repo = MODULE_DIR .. "../../build/libasmtest_drapp." .. ext
  resolved_path = repo
  return repo
end

-- Load the lib defensively: requires DynamoRIO and may be absent, so a failed
-- load is not an error — it just makes available() return false (self-skip).
local L = nil
local ok, lib = pcall(ffi.load, drapp_path())
if ok then L = lib end

local M = {}

-- The absolute path of the libasmtest_drapp this module resolved (env override,
-- else the rock-bundled slot, else the repo build/ tree). Lets a clean-room test
-- assert the bundled tier — not a leaked build/ tree — satisfied the load.
function M.library_path() return resolved_path end

-- The rock-bundled DR client shipped alongside libasmtest_drapp, if present:
-- $ASMTEST_DRCLIENT wins, else native/<os>-<arch>/libasmtest_drclient.<ext> next
-- to this file, else the <repo>/build tree; nil when none exists (the C side then
-- takes its own $ASMTEST_DRCLIENT / repo-build fallback). Mirrors drtrace.py's
-- _default_client.
local function default_client()
  local env = os.getenv("ASMTEST_DRCLIENT")
  if env and env ~= "" then return env end
  local os_name, arch, ext = bundled_slot()
  local cands = {
    MODULE_DIR .. "native/" .. os_name .. "-" .. arch .. "/libasmtest_drclient." .. ext,
    MODULE_DIR .. "../../build/libasmtest_drclient." .. ext,
  }
  for _, c in ipairs(cands) do
    local f = io.open(c, "r")
    if f then f:close(); return c end
  end
  return nil
end

-- ---- Host-native machine code in real executable (W^X) memory ----
local NativeCode = {}
NativeCode.__index = NativeCode
M.NativeCode = NativeCode

-- Materialize a Lua string of host-native machine code into executable memory.
function NativeCode.from_bytes(bytes)
  assert(L, "libasmtest_drapp not loaded")
  local ec = ffi.new("asmtest_exec_code_t[1]")
  local rc = L.asmtest_exec_alloc(ffi.cast("const uint8_t*", bytes), #bytes, ec)
  if rc ~= ASMTEST_DR_OK then
    error("asmtest_exec_alloc failed: " .. tonumber(rc))
  end
  return setmetatable({ ec = ec, freed = false }, NativeCode)
end

-- The executable mapping's base address (offset 0 = entry) as a cdata pointer.
function NativeCode:base() return self.ec[0].base end
-- The number of code bytes.
function NativeCode:length() return tonumber(self.ec[0].len) end
-- Invoke the code through a function pointer, passing two C longs and reading the
-- result back as a long (the SysV integer ABI). Returns a Lua number.
function NativeCode:call(a, b)
  local fn = ffi.cast("long(*)(long,long)", self.ec[0].base)
  return tonumber(fn(a or 0, b or 0))
end
function NativeCode:free()
  if not self.freed then
    L.asmtest_exec_free(self.ec)
    self.freed = true
  end
end

-- ---- An app-owned coverage recorder for a registered native region ----
local NativeTrace = {}
NativeTrace.__index = NativeTrace
M.NativeTrace = NativeTrace

-- ---- process-wide lifecycle ----

-- True if the tier can run (built + libdynamorio resolvable). False on a failed
-- library load OR when asmtest_dr_available() reports 0.
function NativeTrace.available()
  if not L then return false end
  return L.asmtest_dr_available() ~= 0
end

-- Bring DynamoRIO up in-process and take over. `opts` is an optional table:
--   client          path to libasmtest_drclient.so (else $ASMTEST_DRCLIENT)
--   dynamorio_home  DR runtime root (else $ASMTEST_DR_LIB / rpath)
--   client_options  extra client options
--   mode            process-init default recording mode (0 = blocks)
-- error()s on a nonzero return from init or start.
function NativeTrace.initialize(opts)
  assert(L, "libasmtest_drapp not loaded")
  opts = opts or {}
  local o = ffi.new("asmtest_drtrace_options_t")
  -- Default the client to the rock-bundled libasmtest_drclient (honoring
  -- $ASMTEST_DRCLIENT first) when none is passed; a still-nil client_path stays
  -- NULL so the C side takes its own env / repo-build fallback.
  o.client_path = opts.client or default_client()
  o.dynamorio_home = opts.dynamorio_home
  o.client_options = opts.client_options
  o.mode = opts.mode or 0
  local rc = L.asmtest_dr_init(o)
  if rc ~= ASMTEST_DR_OK then error("asmtest_dr_init failed: " .. tonumber(rc)) end
  rc = L.asmtest_dr_start()
  if rc ~= ASMTEST_DR_OK then error("asmtest_dr_start failed: " .. tonumber(rc)) end
end

function NativeTrace.shutdown()
  if L then L.asmtest_dr_shutdown() end
end

-- The exported fixture (a*2+b) the symbol-mode test traces by name. Returns a
-- Lua number.
function M.symbol_demo(a, b)
  assert(L, "libasmtest_drapp not loaded")
  return tonumber(L.asmtest_symbol_demo(a, b))
end

-- Count of illegal marker operations since init; 0 means every marker balanced.
function NativeTrace.marker_error()
  return tonumber(L.asmtest_dr_marker_error())
end

-- ---- per-trace ----

-- Allocate an app-owned trace. `blocks` is the basic-block capacity (default 64);
-- `instructions` the ordered-instruction-stream capacity (default 0 = blocks
-- only). NOTE: the C constructor takes insns FIRST, blocks SECOND.
function NativeTrace.new(blocks, instructions)
  assert(L, "libasmtest_drapp not loaded")
  blocks = blocks or 64
  instructions = instructions or 0
  local h = L.asmtest_trace_new(instructions, blocks)
  if h == nil then error("asmtest_trace_new failed") end
  return setmetatable({ h = h }, NativeTrace)
end

-- Register a non-overlapping native code range under `name`, recording coverage
-- into this trace. error()s on failure.
function NativeTrace:register(name, code)
  local rc = L.asmtest_dr_register_region(name, code:base(), code:length(), self.h)
  if rc ~= ASMTEST_DR_OK then
    error(string.format("register_region(%q) failed: %d", name, tonumber(rc)))
  end
  return self
end

-- Symbol mode: trace a named exported function by NAME, with no begin/end
-- markers — always-on recording for [entry, entry+max_len). error()s on failure.
function NativeTrace:register_symbol(symbol, max_len)
  max_len = max_len or 256
  local rc = L.asmtest_dr_register_symbol(symbol, max_len, self.h)
  if rc ~= ASMTEST_DR_OK then
    error(string.format("register_symbol(%q) failed: %d", symbol, tonumber(rc)))
  end
  return self
end

function NativeTrace:unregister(name)
  L.asmtest_dr_unregister_region(name)
end

-- Run `fn` inside a balanced begin/end marker pair for the named region. The end
-- marker always fires (pcall), even if `fn` errors, so markers stay balanced;
-- the error is then re-raised.
function NativeTrace:region(name, fn)
  L.asmtest_trace_begin(name)
  local ran_ok, err = pcall(fn)
  L.asmtest_trace_end(name)
  if not ran_ok then error(err) end
end

-- True if the basic block at byte-offset `off` (from the region entry) was entered.
function NativeTrace:covered(off)
  return L.asmtest_trace_covered(self.h, off) ~= 0
end

-- Number of distinct basic blocks recorded so far.
function NativeTrace:blocks_len()
  return tonumber(L.asmtest_emu_trace_blocks_len(self.h))
end

-- Total instructions recorded so far (only when allocated with instructions > 0).
function NativeTrace:insns_total()
  return tonumber(L.asmtest_emu_trace_insns_total(self.h))
end

-- The distinct basic-block start offsets recorded, in first-seen order, as a
-- 1-based Lua array of numbers.
function NativeTrace:block_offsets()
  local n = tonumber(L.asmtest_emu_trace_blocks_len(self.h))
  local t = {}
  for i = 0, n - 1 do
    t[i + 1] = tonumber(L.asmtest_emu_trace_block_at(self.h, i))
  end
  return t
end

-- The ordered instruction-offset stream actually stored — each executed
-- instruction's offset in execution order, up to the trace's insns capacity
-- (insns_len, not the possibly-larger insns_total) — as a 1-based Lua array.
function NativeTrace:insn_offsets()
  local n = tonumber(L.asmtest_emu_trace_insns_len(self.h))
  local t = {}
  for i = 0, n - 1 do
    t[i + 1] = tonumber(L.asmtest_emu_trace_insn_at(self.h, i))
  end
  return t
end

function NativeTrace:free()
  if self.h ~= nil then
    L.asmtest_trace_free(self.h)
    self.h = nil
  end
end

return M
